### PR TITLE
Homework UniquePtr implementation

### PR DESCRIPTION
**Pull request by Marcinati**

### CODE REVIEW

File: homework/unique_ptr/UniquePtr.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    }
+    return * this;
+}
+
+template <typename T>
+T & UniquePtr<T>::operator*() const {
+    if (rawPtr_) {
+        return * rawPtr_;
+    } else {
+        throw InvalidDereference("You can't dereference a null pointer.");
```
Comment (MaciejSalamonski) [inline homework/unique_ptr/UniquePtr.hpp:57]: +1

**Review by MaciejSalamonski — APPROVED** [review]
Good job :)

**Review by MaciejSalamonski — APPROVED** [review]
I don't have any comments :) good job

File: homework/unique_ptr/UniquePtr.hpp (hunk 1 above)
Comment (jakjan95) [inline homework/unique_ptr/UniquePtr.hpp:57]: here I would rather put condition !rawPtr_ for throwing exception, and for function by default returning *rawPtr_

File: homework/unique_ptr/UniquePtr.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    T * released = nullptr;
+    std::swap(released, rawPtr_);
+    return released;
+}
+
+template <typename T>
+void UniquePtr<T>::reset(T * newPtr) { 
+    delete rawPtr_;
+    rawPtr_ = newPtr;
+}
```
Comment (jakjan95) [inline homework/unique_ptr/UniquePtr.hpp:82]: here you can put default argument which is nullptr, to allow use just simple reset()

**Review by jakjan95 — COMMENTED** [review]
For me code looks good, i didn't find any magic values in tests, I left some advices for improving code :)

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have nothing to add that was not mentioned by previous reviewers.

File: homework/unique_ptr/UniquePtr.hpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    UniquePtr(const UniquePtr<T> &) = delete;
+    UniquePtr<T> & operator=(const UniquePtr<T> &) = delete;
+
+    UniquePtr<T> & operator=(UniquePtr<T> && otherUniquePtr);
+
+    T & operator*() const;
+    T * operator->() const;
+    T * get() const;
+    T * release();
+    void reset(T * newPtr);
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.hpp:22]: default value would be useful
```suggestion
    void reset(T * newPtr = nullptr);
```

File: homework/unique_ptr/UniquePtr-ut.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    ASSERT_EQ(*uniquePtr, testDistance);
+}
+
+TEST_F(UniquePointerTest, ShouldUseReset) {
+    uniquePtr.reset(new int{newTestDistance});
+    ASSERT_EQ(*uniquePtr, newTestDistance);
+}
+
+TEST_F(UniquePointerTest, ShouldNotDereferenceANullptr) {
+    uniquePtr.reset(nullptr);
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr-ut.cpp:34]:
```suggestion
    uniquePtr.reset();
```
allow calling it without argument

File: homework/unique_ptr/UniquePtr-ut.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    auto rawPtr = uniquePtr.get();
+    ASSERT_EQ(rawPtr, uniquePtr.get());
+}
+
+TEST_F(UniquePointerTest, ShouldCallMethodByArrowOperator) {
+    int otherDistance = 30;
+    int otherTime = 2;
+    UniquePtr<Velocity> VelocityPtr = new Velocity(otherDistance, otherTime);
+    ASSERT_EQ(VelocityPtr->countVelocity(), otherDistance / otherTime);
+}
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr-ut.cpp:59]: missing tests for moving unique_ptrs

**Review by ziobron — CHANGES_REQUESTED** [review]

File: homework/unique_ptr/UniquePtr-ut.cpp (hunk 5 above)
Comment (Marcinati) [reply homework/unique_ptr/UniquePtr-ut.cpp:59]: two tests added

File: homework/unique_ptr/UniquePtr-ut.cpp [hunk 6]
```diff
@@ -52,8 +54,19 @@ TEST_F(UniquePointerTest, ShouldGetAPointer) {
...
+    auto otherVelocityPtr = std::move(VelocityPtr);
+    ASSERT_EQ(VelocityPtr.get(), nullptr);
+    ASSERT_EQ(otherVelocityPtr.get(), nullptr);
+}
+
+TEST_F(UniquePointerTest, SecondUniquePtrShouldGetValuesFromMovedOne) {
+    UniquePtr<Velocity> VelocityPtrFirst = new Velocity(otherDistance, otherTime);
+    UniquePtr<Velocity> VelocityPtrSecond = std::move(VelocityPtrFirst);
+    ASSERT_EQ(VelocityPtrSecond.get()->countVelocity(), otherDistance / otherTime);
+}
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr-ut.cpp:72]: there is still no test for the move assignment operator.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: homework/unique_ptr/UniquePtr-ut.cpp (hunk 6 above)
Comment (Marcinati) [reply homework/unique_ptr/UniquePtr-ut.cpp:73]: sorry for misunderstanding :)

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
36 XP za implementację unique_ptra (3784)
{ "requirement_ids": [3784], "points": [36] }

**Pull request closed without merging**